### ISSUE TITLE
Gridcell water balance errors for some PLUMBER2 sites with branch_tags/PPE.n11_ctsm5.1.dev030

### ISSUE THREAD

**Issue by olyson** [ESCOMP/CTSM#1590]
### Brief summary of bug

Some PLUMBER2 sites throw gridcell water balance errors with branch_tags/PPE.n11_ctsm5.1.dev030.

### General bug information

**CTSM version you are using:** branch_tags/PPE.n11_ctsm5.1.dev030

**Does this bug cause significantly incorrect results in the model's science?** Unknown

**Configurations affected:** Single point PLUMBER2 tower site simulations.  --compset I1PtClm51SpRs --res CLM_USRDAT

### Details of bug

I'm attempting to run PLUMBER2 sites in preparation for developing scripts to run PPE for these sites.  Most of the sites run to completion but several crash with gridcell water balance errors ranging from e-4 to e-2.  The cause of this seems to be that the beginning water balance (begwb_grc) of the problematic time step is different from the ending water balance (endwb_grc) of the previous time step.  The corresponding column water balance checks are fine.  There is only 1 column for the site I'm investigating so there is obviously something different in the terms for the grc and col checks.

### Important details of your setup / configuration so we can reproduce the bug

The setup scripts and input files for these sites were ported from a release branch tag (release-clm5.0.34) version of these simulations.  Other than some modifications to accommodate infrastructure changes, I don't think there is anything uniquely different about these simulations.  I'm curious whether anyone has reported issues with water balance in NEON simulations.
The case directory for the site I'm currently investigating is:

/glade/work/oleson/PPE.n11_ctsm5.1.dev030/cime/scripts/CTL2010SP/basecases/PPEn11_CTL2010SPSpin_AU-Tum

**Comment by ekluzek** [comment]
I did some checking to see if there might have been a fix on CTSM main-dev that would cover this, and I didn't see anything that looks like it. So I don't think that updating to the latest would help the issue. I also searched for other issues like it either open or closed and didn't find anything.

In looking at the case, what happens when you run with irrigate off? And for that matter what happens when irrigate is on with a stub-river model? Do we expect that to function correctly?

**Comment by olyson** [comment]
I didn't see any reference to this type of problem either.
The particular site I'm looking at is 100% BET temperate so I assume irrigate doesn't have an effect.
This seems to be associated with a dribble term at the first of a new year: qflx_liq_dynbal_left_to_dribble.
There is a dribble term that is inherited from the restart file: qflx_liq_dynbal_amt_to_dribble, that is reduced at each time step until it is zero which seems to be the proper behavior.  But then a new dribble amount is calculated and added to begwb_grc.  This amount then disappears (set to zero) in the calculation for endwb_grc.  The difference seems to be the error amount.  I'm looking to see what term(s) enter into the calculation of the new dribble amount.

**Comment by ekluzek** [comment]
Ohhh, nice analysis. Maybe that dribble term should just be eliminated for cases like this? It seems it would be a bit nonsensical anyway. What happens if you zero out that term in the finidat file? Does it work then?

**Comment by wwieder** [comment]
You're amazing @olyson why are you reading in a restart file for these simulations?  Is the restart coming from a global case?  The NEON cases all are initialized as cold starts, so I don't think this is an error we've seen.

**Comment by olyson** [comment]
I haven't tested zeroing out that term but I think that would probably work.
To answer Will's question, yes I'm using the default global restart.  A cold start would probably fix the problem also.

However, the yearfrac variable that is used to determine the amount left to dribble seems to have some strange behavior at the end of the year.
For a time step of 3600s, yearfrac seems to track as expected until the end of the year when it unexpectedly decreases, 

 yearfrac in get_amount_left_to_dribble_beg:   0.999657534246575
 yearfrac in get_amount_left_to_dribble_beg:   0.999771689497717
 yearfrac in get_amount_left_to_dribble_beg:   0.997153916211293
 yearfrac in get_amount_left_to_dribble_beg:   0.000000000000000E+000

i.e., the change from 0.99977 to 0.99715

The cause of that is in the function get_curr_yearfrac in clm_time_manager:

    cday          = get_curr_calday(offset=offset)
    days_per_year = get_days_per_year()
    get_curr_yearfrac = (cday - 1._r8)/days_per_year

This is called by get_prev_yearfrac with an offset of -dtime.
This offset is used to calculate cday.
But the offset is not used to calculate days_per_year.

In this particular case, the simulation is transitioning from a year with 365 days to a year with 366 days.
The fact that the offset is not being used means that days_per_year will be equal to 366 days a time step too early.
I'm wondering if the offset should also be passed to get_days_per_year, i.e.,

    days_per_year = get_days_per_year(offset=offset)

I've done this with some SourceMods and it seems to fix the problem.

**Comment by billsacks** [comment]
Awesome sleuthing @olyson ! Thanks a lot for tracking down the cause of this error. I agree with your proposed fix. I felt this deserved some unit tests, so I have added some on a branch, along with actually making the fix. I have confirmed that one of my new unit tests (`getPrevYearfrac_leapYearAtStartOfYear_returnsCorrectValue`) fails before putting in place the fix and passes after the fix. PR incoming.
